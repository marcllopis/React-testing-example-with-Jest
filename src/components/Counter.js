/* 3rd party imports */
import React, { Component } from 'react';

// initial state is null
// after component didmount is a number
// methods
// functions
// render method renders stuff

export const plusOne = a => a + 1;

export const minusOne = a => a - 1;


class Counter extends Component {

  state = {
    counter: null
  };

  componentDidMount() {
    this.setState({ counter: 0 });
  }

  increment = () =>
    this.setState({ counter: plusOne(this.state.counter) })

  decrement = () => {
    this.setState({ counter: minusOne(this.state.counter) });
  }

  render() {
    return (
      <div>
        <h1>This is the title</h1>
        <h2>{this.state.counter}</h2>
        <button onClick={this.increment}>+</button>
        <button onClick={this.decrement}>-</button>
      </div>
    )
  }
};

export default Counter;
