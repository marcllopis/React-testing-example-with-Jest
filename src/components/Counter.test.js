/* 3rd party imports */
import React from 'react';
import renderer from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import { shallow, configure } from 'enzyme';

import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

import Counter, {minusOne, plusOne} from './Counter';


describe('How to test a component:\n', () => {

  describe('Counter:', () => {

    it('shallow renders properly', () => {
      const wrapper = shallow(<Counter />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('is mounted properly', () => {
      const domTree = renderer.create(<Counter />).toJSON();
      expect(domTree).toMatchSnapshot();
    });

    it('constructor method: it\'s initial counter is null', () => {
      // arrange
      const wrapper = shallow(<Counter />,
        { disableLifecycleMethods: true });
      const expected = null;

      // assert
      expect(wrapper.state('counter')).toEqual(expected);
    });

    it('componentDidMount method: it\'s counter is 0', () => {
      // arrange
      const wrapper = shallow(<Counter />,
        { disableLifecycleMethods: true });
      const initialState = { counter: null };
      const expectedState = { counter: 0 };

      // assert
      expect(wrapper.instance().state).toEqual(initialState);

      // act
      wrapper.instance().componentDidMount();

      // assert
      expect(wrapper.instance().state).toEqual(expectedState);
    });

    it('increment method actually increments the counter value', () => {
      // arrange
      const instance = shallow(<Counter />).instance();
      const initialState = { counter: 43 };
      const expectedState = { counter: 44 };
      instance.setState(initialState);

      // assert
      expect(instance.state).toEqual(initialState);

      // act
      instance.increment();

      // assert
      expect(instance.state).toEqual(expectedState);
    });

    it('decrement method actually decrease the counter value', () => {
      // arrange
      const instance = shallow(<Counter />).instance();
      const initialState = { counter: 43 };
      const expectedState = { counter: 42 };
      instance.setState(initialState);

      // assert
      expect(instance.state).toEqual(initialState);

      // act
      instance.decrement();

      // assert
      expect(instance.state).toEqual(expectedState);
    });
  });

  describe('Helper functions:', () => {
    it('plusOne fn is defined and returns a var incremented', () => {
      expect(plusOne).toBeDefined();
      expect(plusOne(1)).toBe(2);
    });

    it('minusOne fn is defined and returns a var decremented', () => {
      expect(minusOne).toBeDefined();
      expect(minusOne(2)).toBe(1);
    });
  });
});
