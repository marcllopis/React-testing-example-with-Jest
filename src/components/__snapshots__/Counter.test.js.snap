// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`How to test a component:
 Counter: is mounted properly 1`] = `
<div>
  <h1>
    This is the title
  </h1>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    +
  </button>
  <button
    onClick={[Function]}
  >
    -
  </button>
</div>
`;

exports[`How to test a component:
 Counter: shallow renders properly 1`] = `
<div>
  <h1>
    This is the title
  </h1>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    +
  </button>
  <button
    onClick={[Function]}
  >
    -
  </button>
</div>
`;
